.travis.yml

```diff
@@ -7,39 +8,41 @@ matrix:
       env: TOXENV=py27
     - python: '3.4'
       env: TOXENV=py34
+      dist: trusty
     - python: '3.5'
       env: TOXENV=py35
     - python: '3.6'
       env: TOXENV=py36
     - python: '3.7'
       env: TOXENV=py37
-      dist: xenial
-      sudo: true
     - python: 'pypy'
       env: TOXENV=pypy
+      dist: trusty
     - python: 'pypy3'
       env: TOXENV=pypy3
-    - env: TOXENV=flake8
-    - env: TOXENV=checkspelling
-
-addons:
-  apt:
-    packages:
-    - libtidy-0.99-0
-    - aspell
-    - aspell-en
+      dist: trusty
+    - env: TOXENV=flake8,checkspelling
+      addons:
+        apt:
+          packages:
+          - libtidy-0.99-0
+          - aspell
+          - aspell-en
 
 install:
-  - pip install 'setuptools>=36'
-  - pip install tox
-  - pip install 'coverage<4.0'
-  - pip install codecov
+  # NOTE: setuptools needs to be installed explicitly for py34 (trusty).
+  - pip install 'setuptools>=36' tox
 
 script:
   - tox
 
 after_success:
-  - codecov
+  # Upload coverage information for py* tox environments.
+  - |
+    if [ "${TOXENV#py}" != "$TOXENV" ]; then
+      .tox/$TOXENV/bin/coverage xml
+      bash <(curl -s https://codecov.io/bash) -Z -X gcov -X coveragepy -X search -X xcode -X fix -f coverage.xml -F "${TOXENV//-/,}"
+    fi
```

Comment:
Sorry, I'm not a codecov expert, but how does this make thing better?

Reply:
I think the idea is to only call codecov upload on cases that actually have coverage data. With that said, nothing will upload on cases without coverage data anyways, and I don't see the need to use the codecov bash script to do this.

Reply:
1. codecov-bash is more lightweight and meant to be the go-to implementation (less pip-install etc)
2. it will retry on upload failures
3. the condition could be remozed, but the `-Z` will fail the job on failure (which is good I think)

Reply:
From what I understand, the PyPI codecov package also retries on failure and fails the build.  The only other argument is lightweight.  I don't think the python package is heavy either, but yes, you have to install it, but that isn't really a big deal .  I don't really think we are gaining anything with this change.

Reply:
> From what I understand, the PyPI codecov package also retries on failure

Not a few days ago when I looked into it.

> and fails the build

Yes, there is also an option for it.

But OK, it's not critical to have this change.

It's just from my box of things I like to do with Travis configs.

Reply:
> Not a few days ago when I looked into it.

See here: https://github.com/codecov/codecov-python/blob/master/codecov/__init__.py#L765. Not sure how well it works, but it's technically there.  I guess my point is I've never personally felt the `codecov` PyPI package wasn't fitting my needs, but maybe @waylan feels differently.

Reply:
Let's revert this. It is much more complicated and offers no additional value AFAICT.

Reply:
Just for info: it failed here: https://ci.appveyor.com/project/pytestbot/pytest/builds/20372997/job/vku9wlsxtbjwwk0j#L423

```
      _____          _
     / ____|        | |
    | |     ___   __| | ___  ___ _____   __
    | |    / _ \ / _  |/ _ \/ __/ _ \ \ / /
    | |___| (_) | (_| |  __/ (_| (_) \ V /
     \_____\___/ \____|\___|\___\___/ \_/
                                    v2.0.15
==> Detecting CI provider
    AppVeyor Detected
    Fixing merge commit SHA
==> Preparing upload
    Error running `cd 'C:\projects\pytest' && git ls-files`: None
XX> Skip processing gcov
==> Collecting reports
XX> Searching for reports disabled
    Targeting specific files
    + coverage.xml bytes=1071184
==> Appending environment variables
    + TOXENV
==> Uploading
    .url https://codecov.io
    .query commit=3acfa3abdc7577f97bb40b9790c10d43c0e486e4&branch=master&job=pytestbot%2Fpytest%2F1.0.5975&pr=4409&service=appveyor&build=vku9wlsxtbjwwk0j&slug=pytest-dev%2Fpytest&flags=py27%2Cwindows&package=py2.0.15
    Pinging Codecov...
Error: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))
Tip: See all example repositories: https://github.com/codecov?query=example
Support channels:
  Email:   hello@codecov.io
  IRC:     #codecov
  Gitter:  https://gitter.im/codecov/support
  Twitter: @codecov
```